src/base/IoManip.h

```diff
@@ -220,5 +220,48 @@ operator <<(std::ostream &os, const AsList<Container> &manipulator)
 template <typename Container>
 inline auto asList(const Container &c) { return AsList<Container>(c); }
 
+/// Helps print T object at most once per AtMostOnce<T> object lifetime.
+/// T objects are printed to std::ostream using operator "<<".
+///
+/// \code
+/// auto headerOnce = AtMostOnce("Transaction Details:\n");
+/// if (detailOne)
+///     os << headerOnce << *detailOne;
+/// if (const auto detailTwo = findAnotherDetail())
+///     os << headerOnce << *detailTwo;
+/// \endcode
+template <class T>
+class AtMostOnce
+{
+public:
+    /// caller must ensure `t` lifetime extends to the last use of this AtMostOnce instance
+    explicit AtMostOnce(const T &t): toPrint(t) {}
+
+    void print(std::ostream &os) {
+        if (!printed) {
+            os << toPrint;
+            printed = true;
+        }
+    }
+
+private:
+    const T &toPrint;
+    bool printed = false;
+};
+
+/// Prints AtMostOnce argument if needed. The argument is not constant to
+/// prevent wrong usage:
+///
+/// \code
+/// /* Compiler error: cannot bind non-const lvalue reference to an rvalue */
+/// os << AtMostOnce(x);
+/// \endcode
+template <class T>
+inline auto &
+operator <<(std::ostream &os, AtMostOnce<T> &a) {
```

Comment:
> > Amos: I still predict the same issue when this manipulator is used like all the other stream manipulators. Specifically using the pattern: `... << AtMostOnce("foo") << ...`
> 
> Alex: That usage will be a bug

Fortunately, it will be a compile-time bug. In other words, code like `<< AtMostOnce("foo") <<` does not compile. FWIW, I was kind of hoping for that outcome, but did not want to claim that until I actually tested PR code. I have now adjusted AtMostOnce documentation to reflect that protection.